src/consul.rs

```diff
@@ -0,0 +1,399 @@
+        let session_response: CreateSessionResponse = resp.json().await?;
+        Ok(session_response.id)
+    }
+
+    /// Renew the Consul session
+    pub async fn renew_session(&self, session_id: Uuid) -> Result<(), anyhow::Error> {
+        let session_url = self
+            .session_api_base_url
+            .join(&format!("renew/{}", session_id))?;
+
+        let mut req = self.http_client.put(session_url);
+
+        // Set dc if it is provided in the config
+        if let Some(dc) = &self.datacenter {
+            req = req.query(&[("dc", dc)]);
+        }
+
+        req.send().await?.error_for_status()?;
+        println!("Renewed Consul session: {}", session_id);
+        Ok(())
+    }
+
+    /// Acquire a lock
+    ///
+    /// Times out after a while and returns an error if it does.
+    pub async fn acquire_lock(&self, session_id: Uuid) -> Result<()> {
+        let mut interval = interval(Duration::from_secs(10));
+        interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+
+        loop {
+            let lock_url = self.kv_api_base_url.join(CONSUL_STORE_KEY)?;
+
+            let mut req = self.http_client.put(lock_url);
+            req = req.query(&[("acquire", &session_id.to_string())]);
+
+            // Set dc if it is provided in the config
+            if let Some(dc) = &self.datacenter {
+                req = req.query(&[("dc", dc)]);
+            }
+
+            let resp = req.send().await?;
+            let body = resp.text().await?;
+
+            // If the lock is acquired, the response body will be "true"
+            if body.starts_with("true") {
+                println!("=====> Acquired Consul lock");
+                return Ok(());
+            }
+
+            println!("=====> Failed to acquire Consul lock");
+            // We limit re-checks to at most every 10 seconds, so we don't spam the server in case we
+            // didn't acquire the lock even though it claims it to be free.
+            interval.tick().await;
+            // Wait for lock to be free
+            self.wait_for_lock().await?;
+        }
+    }
+
+    async fn wait_for_lock(&self) -> Result<(), anyhow::Error> {
+        let mut consul_index: Option<u64> = None;
+        loop {
+            let lock_url = self.kv_api_base_url.join(CONSUL_STORE_KEY)?;
+
+            let mut req = self.http_client.get(lock_url);
+
+            // Set dc if it is provided in the config
+            if let Some(dc) = &self.datacenter {
+                req = req.query(&[("dc", dc)]);
+            }
+
+            if let Some(index) = consul_index.take() {
+                req = req.query(&[("index", &index.to_string())]);
+            }
+
+            let response = req.send().await?;
+
+            consul_index = response
+                .headers()
+                .get("X-Consul-Index")
+                .and_then(|value| value.to_str().ok())
+                .and_then(|value| value.parse().ok());
```

Comment:
As per https://github.com/codercengiz/consul-external-dns/pull/1#discussion_r1589324284 we cannot rely on this being a `u64`, we have to leave it as the raw string we received.

Reply:
Fixed